### PR TITLE
[BEAM-13888] Add unit testing to ioutilx

### PR DESCRIPTION
**Pull request by Abacn**
Add unit testing for everything in ioutilx package

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment ( @jrmccluskey ).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	r := strings.NewReader(testString)
+
+	// test good read
+	data, err := ReadN(r, testLength)
+	// err is expected to be nil
+	if err != nil {
+		t.Errorf("Failed to read data: %v", err)
+	}
+	// length of data is expected to match testLength
+	if (len(data) != testLength) {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:36]: No need to use the parentheses here, and you can use local variable assignment within the if statement to make your error message formatting more concise:
```suggestion
	if got, want := len(data),  testLength; got != want {
	     t.Errorf("got length %v, want %v", got, want)
	}
```

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		t.Errorf("Got %v, wanted %v", len(data), testLength)
+	}
+	// content of data is expected to match testString
+	if (string(data) != testString) {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:40]: Same pattern with the parentheses and "got, want" style here

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		t.Errorf("Failed to read data: %v", err)
+	}
+	// content of data is expected to match testString
+	if (string(data[:]) != testString) {
+		t.Errorf("Got %q, wanted %q", string(data[:]), testString)
+	}
+}
+
+func TestReadUnsafe(t *testing.T) {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:69]: Same style comments as above

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	// test bad read
+	data, err = ReadN(r, testLength+1)
+	// err is expected to be io.EOF
+	if err != io.EOF {
+		t.Errorf("Got %v, wanted %v", err, io.EOF)
+	}
+}
+
+func TestReadNBufUnsafe(t *testing.T) {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:52]: Same style comments as above

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	}
+}
+
+func TestReadUnsafe(t *testing.T) {
+	const testString = "hello world!"
+	const testLength = len(testString)
+	r := strings.NewReader(testString)
+	var data [testLength]byte
+
+	length, err := ReadUnsafe(r, data[:])
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:75]: You should be able to pass `data` here without specifying the range for the slice. 

```suggestion
	length, err := ReadUnsafe(r, data)
```

File: sdks/go/pkg/beam/core/util/ioutilx/write_test.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+package ioutilx
+
+import (
+	"bufio"
+	"bytes"
+	"testing"
+)
+
+func TestWriteUnsafe(t *testing.T) {
+	var buff bytes.Buffer
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/write_test.go:25]: Nit: Go style usually calls this `buf`

File: sdks/go/pkg/beam/core/util/ioutilx/write_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+package ioutilx
+
+import (
+	"bufio"
+	"bytes"
+	"testing"
+)
+
+func TestWriteUnsafe(t *testing.T) {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/write_test.go:24]: Same style comments as in write.

**Review by jrmccluskey — CHANGES_REQUESTED** [review]
Good start! Have some style changes for you to make and a few things that leverage Go features.

**Comment by jrmccluskey** [comment]
Also: fix the typo in the PR title so it gets mapped to the Jira ticket properly, and when you request a reviewer you do it in a follow-up comment and not the body of the PR template.

**Comment by Abacn** [comment]
@jrmccluskey Thanks for feedbacks! Comments addressed.

File: sdks/go/pkg/beam/core/util/ioutilx/read_test.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	r := strings.NewReader(testString)
+	data := make([]byte, testLength)
+
+	err := ReadNBufUnsafe(r, data)
+	// err is expected to be nil
+	if err != nil {
+		t.Errorf("failed to read data: got %v", err)
+	}
+	// content of data is expected to match testString
+	if got, want := string(data[:]), testString; got != want {
```
Comment (jrmccluskey) [inline sdks/go/pkg/beam/core/util/ioutilx/read_test.go:70]:
```suggestion
	if got, want := string(data), testString; got != want {
```

**Review by jrmccluskey — APPROVED** [review]
LGTM, going to add a committer so the workflows can run and we can get this merged in!

**Comment by Abacn** [comment]
Sorry, missed a few style change request and fixed.

**Comment by jrmccluskey** [comment]
R: @youngoli

**Comment by johnjcasey** [comment]
LGTM

**Review by youngoli — CHANGES_REQUESTED** [review]
Looks really good. I'm being a bit more nitpicky than I usually would be just to explain Go conventions, but even then this is nearly there.

**Comment by Abacn** [comment]
@youngoli Thanks for feedbacks! Comments addressed.

**Review by youngoli — APPROVED** [review]
Looks good to me!

**Pull request merged**